eyeball-im-util/src/vector/limit.rs

```diff
@@ -0,0 +1,372 @@
+    /// The limit is dynamic, i.e. it changes over time based on values that are polled
+    /// from another `Stream` (ref. [`Self::limit_stream`]).
+    ///
+    /// Because the limit is dynamic, an internal buffered vector is kept, so that
+    /// the adapter knows which values can be added when the limit is increased, or
+    /// when values are removed and new values must be inserted. This fact is important
+    /// if the items of the `Vector` have a non-negligeable size.
+    ///
+    /// It's OK to have a limit larger than the length of the observed `Vector`.
+    ///
+    /// Note that this adapter won't produce anything until a new limit is polled.
+    #[project = DynamicLimitProj]
+    pub struct DynamicLimit<S, L>
+    where
+        S: Stream,
+        S::Item: VectorDiffContainer,
+        L: Stream<Item = usize>,
+    {
+        // The main stream where items are polled from.
+        #[pin]
+        inner_stream: S,
+
+        // The limit stream where new limits are polled from.
+        #[pin]
+        limit_stream: L,
+
+        // The buffered vector that are built from the main stream's items. It's
+        // used to provide missing items e.g. when the limit increases.
+        buffered_vector: Vector<VectorDiffContainerStreamElement<S>>,
+
+        // The current limit.
+        limit: usize,
+
+        // This adapter is not a basic filter: It can produce items. For example, if the
+        // vector is [10, 11, 12, 13], with a limit of 2; then if an item is poped
+        // front, 10 is removed, but 12 is pushed back as it “enters” the “view”. That's
+        // 2 items to produce. This field contains all items that must be polled before
+        // anything.
+        ready_values: VecDeque<S::Item>,
+    }
+}
+
+impl<S, L> DynamicLimit<S, L>
+where
+    S: Stream,
+    S::Item: VectorDiffContainer,
+    VectorDiffContainerStreamElement<S>: Clone + Send + Sync + 'static,
+    VectorDiffContainerStreamFamily<S>:
+        VectorDiffContainerFamily<Member<VectorDiffContainerStreamElement<S>> = S::Item>,
+    L: Stream<Item = usize>,
+{
+    /// Create a new [`DynamicLimit`] with the given (unlimited) initial values,
+    /// stream of `VectorDiff` updates for those values, and a stream of
+    /// limits.
+    pub fn new(
+        initial_values: Vector<VectorDiffContainerStreamElement<S>>,
+        inner_stream: S,
+        limit_stream: L,
+    ) -> Self {
+        Self {
+            inner_stream,
+            limit_stream,
+            buffered_vector: initial_values,
+            limit: 0,
+            ready_values: VecDeque::new(),
+        }
+    }
+}
+
+impl<S, L> Stream for DynamicLimit<S, L>
+where
+    S: Stream,
+    S::Item: VectorDiffContainer,
+    VectorDiffContainerStreamElement<S>: Clone + Send + Sync + 'static,
+    VectorDiffContainerStreamFamily<S>:
+        VectorDiffContainerFamily<Member<VectorDiffContainerStreamElement<S>> = S::Item>,
```

Comment:
This bound is only fulfilled by the unbatched vector subscriber. It has to be adjusted to work for both (removing the bound), or a second impl has to be added (not sure if that will work, I think Rust will see a conflicting implementation).

Reply:
I've removed this bound. Everything seems to compile just find with this with the existing tests. It was necessary before, but not anymore.